tests/cpp/common/test_host_device_vector.cu

```diff
@@ -213,36 +213,30 @@ TEST(HostDeviceVector, Reshard_PreserveGPUData) {
   auto span = vec.DeviceSpan(0);  // sync to device
   PlusOne(&vec);
 
-  vec.Reshard(GPUDistribution::Empty(), GPUDataPreservation::kPreserve);
+  // GPU data is preserved.
+  vec.Reshard(GPUDistribution::Empty());
   ASSERT_EQ(vec.Size(), h_vec.size());
   ASSERT_TRUE(vec.Devices().IsEmpty());
 
   auto h_vec_1 = vec.HostVector();
   for (size_t i = 0; i < h_vec_1.size(); ++i) {
     ASSERT_EQ(h_vec_1.at(i), i + 1);
   }
-}
-
-TEST(HostDeviceVector, Reshard_DiscardGPUData) {
-  std::vector<int> h_vec (2345);
-  for (size_t i = 0; i < h_vec.size(); ++i) {
-    h_vec[i] = i;
-  }
-  HostDeviceVector<int> vec (h_vec);
-  auto devices = GPUSet::Range(0, 1);
 
   vec.Shard(devices);
-  ASSERT_EQ(vec.DeviceSize(0), h_vec.size());
-  ASSERT_EQ(vec.Size(), h_vec.size());
-  auto span = vec.DeviceSpan(0);  // sync to device
+  span = vec.DeviceSpan(0);  // sync to device
   PlusOne(&vec);
 
-  vec.Reshard(GPUDistribution::Empty(), GPUDataPreservation::kDiscard);
+  // GPU data is discarded.
+  vec.Reshard(GPUDistribution::Empty(), /*preserve_gpu_data=*/false);
   ASSERT_EQ(vec.Size(), h_vec.size());
   ASSERT_TRUE(vec.Devices().IsEmpty());
 
-  auto h_vec_1 = vec.HostVector();
-  ASSERT_TRUE(std::equal(h_vec_1.cbegin(), h_vec_1.cend(), h_vec.cbegin()));
+  auto h_vec_2 = vec.HostVector();
```

Comment:
OK, I think there is some misunderstanding here. By not preserving data, we are only trying to avoid copying the data back from device to host. Host data should be preserved, and then resharded. This is useful for `SparsePage`s where we don't care about the data left on device, but just want to redistribute the host data. Perhaps we should keep the bool as `preserve_gpu_data`?

Reply:
As discussed offline, not preserving data doesn't make much sense at this point, so removed the parameter.